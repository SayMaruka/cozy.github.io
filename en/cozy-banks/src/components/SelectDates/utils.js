const rangedSome = (arr, predicate, start, end) => {
  start = Math.max(start, 0)
  end = Math.min(end, arr.length)
  const step = start > end ? -1 : 1
  if (!arr || arr.length === 0) {
    return false
  }
  for (let i = start; i !== end; i = i + step) {
    if (predicate(arr[i])) {
      return true
    }
  }
  return false
}

export { rangedSome }
