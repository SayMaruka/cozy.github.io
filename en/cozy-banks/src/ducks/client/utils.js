/**
 * All this file should be moved to cozy-client
 */

export const isCollectionLoading = col => {
  if (!col) {
    console.warn('isCollectionLoading called on falsy value.') // eslint-disable-line no-console
    return false
  }
  return col.fetchStatus === 'loading' || col.fetchStatus === 'pending'
}

export const hasBeenLoaded = col => {
  return col.lastFetch
}

export const associateDocuments = (
  originalDocument,
  associationName,
  associationDoctype,
  documentsToAssociate
) => {
  const ids = documentsToAssociate.map(doc => doc.id)

  originalDocument[associationName] = {
    data: documentsToAssociate,
    doctype: associationDoctype,
    name: associationName,
    raw: ids,
    target: {
      ...originalDocument,
      [associationName]: ids
    }
  }

  return originalDocument
}

export const isRevoked = async client => {
  try {
    await client.stackClient.fetchInformation()
    return false
  } catch (err) {
    if (err.message && err.message.indexOf('Client not found') > -1) {
      return true
    } else {
      return false
    }
  }
}

export const checkForRevocation = async client => {
  const revoked = await isRevoked(client)
  if (revoked) {
    client.stackClient.unregister()
    client.handleRevocationChange(true)
  }
}

/**
 * Updates the device notification token
 *
 * @param  {CozyClient} client
 * @param  {String} token  - Device token, from Cordova Push Plugin
 */
export const updateNotificationToken = (client, token) => {
  // Updates local and remote information
  const clientInfos = client.stackClient.oauthOptions
  client.stackClient.updateInformation({
    ...clientInfos,
    notificationDeviceToken: token
  })
}

/**
 * Get current device notification token
 *
 * @param  {CozyClient} client
 * @param  {String} token  - Device token, from Cordova Push Plugin
 */
export const getNotificationToken = client => {
  return client.stackClient.oauthOptions.notification_device_token
}

export const getDocumentIdentity = doc =>
  doc
    ? {
        _id: doc._id,
        _type: doc._type
      }
    : null
