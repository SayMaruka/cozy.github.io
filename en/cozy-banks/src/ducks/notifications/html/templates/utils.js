const Handlebars = require('handlebars')
const overEvery = require('lodash/overEvery')
const fromPairs = require('lodash/fromPairs')
const layouts = require('handlebars-layouts')

layouts.register(Handlebars)

const isOfType = type => node => {
  return node.type == type
}

const isPath = path => node => {
  return node.path.original == path
}

const isExtendBlock = overEvery([isOfType('BlockStatement'), isPath('extend')])
const isContentBlock = overEvery([
  isOfType('BlockStatement'),
  isPath('content')
])

/**
 * Extracts the name of extending template.
 * Extracts the AST of each content block.
 */
const extractInfo = templateContent => {
  const ast = Handlebars.parse(templateContent)
  const extendNode = ast.body.find(isExtendBlock)
  const contentNodes = extendNode.program.body.filter(isContentBlock)

  // Mapping from content block name to content block AST
  const contents = fromPairs(
    contentNodes.map(node => [node.params[0].value, node.program])
  )
  return {
    parentTemplate: extendNode.params[0].value,
    parts: contents,
    ast
  }
}

/**
 * Is used to replace parts of a template by rendered content.
 *
 * @param {Object} blocks - Block name -> Block AST
 * @param {Object} replacements - Block name -> Rendered content
 *
 * @example
 * ```
 * {{#content "emailSubtitle"}}
 *   {{#each transactions as |transaction| }}
 *     <div>{{ transaction.label }}</div>
 *   {{/each}}
 * {{/content}}
 *
 * becomes
 *
 * {{#content "emailSubtitle"}}
 *   <div>{{ transaction.label }}</div>
 *   <div>{{ transaction.label }}</div>
 * {{/content}}
 * ```
 */
const replaceParts = (blocks, replacements) => {
  for (const [blockName, replacement] of Object.entries(replacements)) {
    blocks[blockName].body = [
      {
        type: 'ContentStatement',
        original: replacement,
        value: replacement
      }
    ]
  }
}

module.exports = {
  extractInfo,
  replaceParts
}
