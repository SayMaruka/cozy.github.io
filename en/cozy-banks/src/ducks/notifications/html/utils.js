import cheerio from 'cheerio'
import groupBy from 'lodash/groupBy'
import sortBy from 'lodash/sortBy'
import toPairs from 'lodash/toPairs'
import flow from 'lodash/flow'
import unique from 'lodash/uniq'
import get from 'lodash/get'
import { getDate } from 'ducks/transactions/helpers'
import { mjml2html } from 'mjml'

export const toText = (cozyHTMLEmail, getContent) => {
  const $ = cheerio.load(cozyHTMLEmail)
  const title = $('.header__title')
    .text()
    .trim()
  const descTitle = $('.header__desc__title')
    .text()
    .trim()
  const descSubtitle = $('.header__desc__subtitle')
    .text()
    .trim()
  return `
# Cozy - ${title}
## ${descTitle} - ${descSubtitle}

---------
${getContent($)}
`
}

export const prepareTransactions = function(transactions) {
  const byAccounts = groupBy(transactions, tr => tr.account)

  const groupAndSortByDate = flow(
    transactions => groupBy(transactions, getDate),
    toPairs,
    dt => sortBy(dt, ([date]) => date).reverse()
  )
  Object.keys(byAccounts).forEach(account => {
    byAccounts[account] = groupAndSortByDate(byAccounts[account])
  })

  return byAccounts
}

const billIdFromReimbursement = reimbursement => {
  return reimbursement.billId && reimbursement.billId.split(':')[1]
}

export const treatedByFormat = function(reimbursements, billsById) {
  return unique(
    (reimbursements || [])
      .map(reimbursement => {
        const billId = billIdFromReimbursement(reimbursement)
        return get(billsById, billId + '.vendor')
      })
      .filter(x => x && x !== '')
  ).join(', ')
}

export const renderMJML = mjmlContent => {
  const obj = mjml2html(mjmlContent)
  obj.errors.forEach(err => {
    // eslint-disable-next-line no-console
    console.warn(err.formattedMessage)
  })

  if (obj.html) {
    return obj.html
  } else {
    throw new Error('Error during HTML generation')
  }
}

export const getCurrentDate = () => {
  return new Date()
}
