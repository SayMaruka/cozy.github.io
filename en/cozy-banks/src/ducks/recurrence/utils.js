import startCase from 'lodash/startCase'
import maxBy from 'lodash/maxBy'
import groupBy from 'lodash/groupBy'
import {
  getCategoryId,
  getLabel as getTransactionLabel
} from 'ducks/transactions/helpers'

const RECURRENCE_DOCTYPE = 'io.cozy.bank.recurrence'

export const prettyLabel = label => {
  return label ? startCase(label.toLowerCase()) : ''
}

const mostFrequent = (iterable, fn) => {
  const groups = groupBy(iterable, fn)
  return maxBy(Object.entries(groups), ([, ops]) => ops.length)[0]
}

export const getAutomaticLabelFromBundle = bundle =>
  mostFrequent(bundle.ops, op => op.label)

export const getLabel = bundle => {
  if (bundle.manualLabel) {
    return bundle.manualLabel
  } else {
    return prettyLabel(bundle.automaticLabel)
  }
}

export const getCategories = bundle => {
  return bundle.categoryIds
}

export const getFrequencyWord = freq => {
  if (freq > 5 && freq <= 8) {
    return 'weekly'
  } else if (freq > 26 && freq <= 40) {
    return 'monthly'
  } else if (freq > 300 && freq <= 400) {
    return 'yearly'
  }
}

export const getFrequency = recurrence => {
  return recurrence.stats.deltas.median
}

export const getFrequencyText = (t, recurrence) => {
  const frequency = Math.floor(getFrequency(recurrence))
  const freqWord = getFrequencyWord(frequency)
  const text = freqWord
    ? t(`Recurrence.freq-info-word.${freqWord}`)
    : t(`Recurrence.freq-info`, { frequency })
  return text
}

export const getAmount = bundle => {
  return bundle.amounts[0]
}

export const getCurrency = () => {
  return 'â‚¬'
}

export const makeRecurrenceFromTransaction = transaction => {
  return {
    _type: RECURRENCE_DOCTYPE,
    automaticLabel: getTransactionLabel(transaction),
    stats: {
      deltas: {
        median: 30
      }
    },
    amounts: [transaction.amount],
    categoryIds: [getCategoryId(transaction)]
  }
}
