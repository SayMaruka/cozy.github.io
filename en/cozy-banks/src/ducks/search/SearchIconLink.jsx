import React from 'react'
import Icon from 'cozy-ui/transpiled/react/Icon'
import useBreakpoints from 'cozy-ui/transpiled/react/hooks/useBreakpoints'
import cx from 'classnames'
import styles from './SearchIconLink.styl'
import { BarRight } from 'components/Bar'

import MagnifierIcon from 'cozy-ui/transpiled/react/Icons/Magnifier'

const SearchIconLink = ({ className }) => {
  return (
    <a className={cx(styles.Icon, className)} href="#/search">
      <Icon icon={MagnifierIcon} />
    </a>
  )
}

export const MobileBarSearchIconLink = () => {
  const { isMobile } = useBreakpoints()
  return isMobile ? (
    <BarRight>
      <SearchIconLink />
    </BarRight>
  ) : null
}

export default SearchIconLink
