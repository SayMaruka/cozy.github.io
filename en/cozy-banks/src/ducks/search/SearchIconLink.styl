.Icon
    width 3rem
    height 3rem
    display flex
    justify-content center
    align-items center
    color var(--primaryContrastTextColor)
