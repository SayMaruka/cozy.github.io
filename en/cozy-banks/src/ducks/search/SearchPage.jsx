import React, { useState } from 'react'
import { withRouter } from 'react-router-dom'

import { useQuery } from 'cozy-client'
import { Input, Stack, useI18n, Empty } from 'cozy-ui/transpiled/react'

import { transactionsConn } from 'doctypes'
import { TransactionList } from 'ducks/transactions/Transactions'
import BarTheme from 'ducks/bar/BarTheme'
import TransactionTableHead from 'ducks/transactions/header/TableHead'

import Header from 'components/Header'
import Padded from 'components/Spacing/Padded'
import { PageTitle } from 'components/Title'
import BackButton from 'components/BackButton'

const makeSearch = searchStr => op => {
  return op.label.toLowerCase().includes(searchStr.toLowerCase())
}

const isSearchSufficient = searchStr => searchStr.length > 3

const SearchPage = ({ router }) => {
  const { t } = useI18n()
  const [search, setSearch] = useState(router.params.search || '')
  const handleChange = ev => {
    setSearch(ev.target.value)
  }

  let { data: transactions } = useQuery(
    transactionsConn.query,
    transactionsConn
  )

  transactions = transactions || []

  const searchSufficient = isSearchSufficient(search)
  if (searchSufficient) {
    transactions = transactions.filter(makeSearch(search))
  }

  return (
    <div>
      <BarTheme theme="primary" />
      <Header theme="inverted" fixed>
        <Padded>
          <Stack spacing="l">
            <div>
              <BackButton to="/balances" arrow={true} />
              <PageTitle className="u-lh-tiny">{t('Search.title')}</PageTitle>
            </div>
            <Input
              placeholder={t('Search.input-placeholder')}
              type="text"
              value={search}
              onChange={handleChange}
            />
          </Stack>
        </Padded>
        <TransactionTableHead isSubcategory={false} />
      </Header>
      {!searchSufficient ? <Padded>{t('Search.type-a-search')}</Padded> : null}
      <div className={`js-scrolling-element`}>
        {searchSufficient ? (
          transactions.length > 0 ? (
            <TransactionList transactions={transactions} />
          ) : (
            <Empty
              className="u-mt-large"
              title={t('Search.no-transactions-found', { search })}
              icon={''}
            />
          )
        ) : null}
      </div>
    </div>
  )
}

export default withRouter(SearchPage)
