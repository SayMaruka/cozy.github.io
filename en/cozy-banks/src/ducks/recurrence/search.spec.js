import sortBy from 'lodash/sortBy'
import minBy from 'lodash/minBy'
import maxBy from 'lodash/maxBy'
import uniqBy from 'lodash/uniqBy'
import isArray from 'lodash/isArray'

import { TRANSACTION_DOCTYPE } from 'doctypes'
import { td, tr } from 'txt-table-utils'

import { findAndUpdateRecurrences } from './search'
import fixtures from './fixtures'
import { getLabel, getAmount } from './utils'

const formatBundleExtent = bundle => {
  const oldestOp = minBy(bundle.ops, x => x.date)
  const latestOp = maxBy(bundle.ops, x => x.date)
  return `from ${oldestOp.date.slice(0, 10)} to ${latestOp.date.slice(0, 10)}`
}

const assert = (cond, error) => {
  if (!cond) {
    throw new Error(error)
  }
}

const assertUniqueOperations = recurrence => {
  const uniquedOps = uniqBy(recurrence.ops, op => op._id)
  assert(
    recurrence.ops.length === uniquedOps.length,
    `Duplicate ops in recurrence ${getLabel(recurrence)}`
  )
}

const assertValidRecurrence = bundle => {
  assertUniqueOperations(bundle)
  assert(isArray(bundle.amounts), 'Bundle should have amounts array')
  assert(isArray(bundle.categoryIds), 'Bundle should have categoryIds array')
}

const formatRecurrence = bundle =>
  tr(
    td(getLabel(bundle), 50),
    td(getAmount(bundle), 10, 'right'),
    td(`${bundle.ops.length} operations`, 17, 'right'),
    td(formatBundleExtent(bundle), 30, 'right'),
    td(bundle.stats.deltas.median, 7, 'right')
  )

describe('recurrence bundles', () => {
  it('should find new bundles', () => {
    const transactions = fixtures[TRANSACTION_DOCTYPE]
    const recurrences = []
    const updatedRecurrences = findAndUpdateRecurrences(
      recurrences,
      transactions
    )

    updatedRecurrences.forEach(assertValidRecurrence)
    // eslint-disable
    expect(sortBy(updatedRecurrences.map(formatRecurrence))).toEqual([
      'Cozy Cloud                                         |       3963 |      3 operations |  from 2018-06-01 to 2018-09-28 |    59.5',
      'Echeance Pret 03103 61893938                       |      -1281 |     11 operations |  from 2019-03-05 to 2020-01-06 |      31',
      'Edf Clients Particuliers                           |        -36 |     11 operations |  from 2018-08-24 to 2019-06-24 |      31',
      'Free Mobile                                        |        -17 |     15 operations |  from 2018-05-22 to 2020-01-20 |      30',
      'Free Telecom                                       |        -31 |     20 operations |  from 2018-05-07 to 2020-02-06 |      30',
      'Navigo Annuel Gie Comutitres                       |        -75 |     20 operations |  from 2018-06-04 to 2020-03-03 |      31',
      'Paypal Europe S A R L Et Cie S C A Ech             |        -13 |      4 operations |  from 2018-05-29 to 2018-10-30 |      32',
      'Vr Permanent Bourso                                |        -60 |     22 operations |  from 2018-06-04 to 2020-03-02 |      31',
      'Vr Permanent Loyer Douillet                        |       -913 |     11 operations |  from 2018-06-04 to 2019-04-03 |      31'
    ])
  })

  it('should update existing bundles', () => {
    const transactions = fixtures[TRANSACTION_DOCTYPE].filter(
      tr => tr.date < '2019-05'
    )
    const newTransactions = fixtures[TRANSACTION_DOCTYPE].filter(
      tr => tr.date > '2019-05'
    )
    const recurrences = findAndUpdateRecurrences([], transactions)

    expect(sortBy(recurrences.map(formatRecurrence))).toEqual([
      'Cozy Cloud                                         |       3963 |      3 operations |  from 2018-06-01 to 2018-09-28 |    59.5',
      'Edf Clients Particuliers                           |        -36 |      9 operations |  from 2018-08-24 to 2019-04-24 |      31',
      'Free Telecom                                       |        -31 |     12 operations |  from 2018-05-07 to 2019-04-05 |      30',
      'Navigo Annuel Gie Comutitres                       |        -75 |     10 operations |  from 2018-06-04 to 2019-04-03 |      31',
      'Paypal Europe S A R L Et Cie S C A Ech             |        -13 |      4 operations |  from 2018-05-29 to 2018-10-30 |      32',
      'Vr Permanent Bourso                                |        -60 |     11 operations |  from 2018-06-04 to 2019-04-02 |    30.5',
      'Vr Permanent Loyer Douillet                        |       -913 |     11 operations |  from 2018-06-04 to 2019-04-03 |      31'
    ])

    const updatedRecurrences = findAndUpdateRecurrences(
      recurrences,
      newTransactions
    )

    updatedRecurrences.forEach(assertValidRecurrence)

    expect(sortBy(updatedRecurrences.map(formatRecurrence))).toEqual([
      'Cozy Cloud                                         |       3963 |     13 operations |  from 2018-06-01 to 2020-02-28 |    31.5',
      //      'Digitick Billet Marseille | -42 | 5 operations | from 2019-05-20 to 2020-03-31 | 0',
      //      'Echeance Pret 03103 61893938 | -1281 | 11 operations | from 2019-03-05 to 2020-01-06 | 31',
      'Edf Clients Particuliers                           |        -36 |     11 operations |  from 2018-08-24 to 2019-06-24 |      31',
      //      'Free Mobile | -17 | 15 operations | from 2018-05-22 to 2020-01-20 | 30',
      'Free Telecom                                       |        -31 |     25 operations |  from 2018-05-07 to 2020-03-06 |      30',
      'Navigo Annuel Gie Comutitres                       |        -75 |     20 operations |  from 2018-06-04 to 2020-03-03 |      31',
      'Paypal Europe S A R L Et Cie S C A Ech             |        -13 |      8 operations |  from 2018-05-29 to 2019-06-11 |      20',
      'Vr Permanent Bourso                                |        -60 |     22 operations |  from 2018-06-04 to 2020-03-02 |      31',
      'Vr Permanent Loyer Douillet                        |       -913 |     11 operations |  from 2018-06-04 to 2019-04-03 |    30.5'
    ])
  })
})
