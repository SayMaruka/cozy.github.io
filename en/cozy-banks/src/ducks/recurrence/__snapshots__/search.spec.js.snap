// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`recurrence bundles should find new bundles (split brand necessary) 1`] = `
"75 Netflix Com                                     |        -11 |     10 operations |  from 2020-09-01 to 2021-06-01 |      30
Spotify P 127537 A 66                              |        -14 |      7 operations |  from 2020-10-01 to 2021-04-01 |      30"
`;

exports[`recurrence bundles should find new bundles 1`] = `
"Cozy Cloud                                         |       3963 |      3 operations |  from 2018-06-01 to 2018-09-28 |    59.5
Echeance Pret 03103 61893938                       |      -1281 |     11 operations |  from 2019-03-05 to 2020-01-06 |      31
Edf Clients Particuliers                           |        -36 |     11 operations |  from 2018-08-24 to 2019-06-24 |      31
Free Mobile                                        |        -17 |     15 operations |  from 2018-05-22 to 2020-01-20 |      30
Free Telecom                                       |        -31 |     20 operations |  from 2018-05-07 to 2020-02-06 |      30
Navigo Annuel Gie Comutitres                       |        -75 |     20 operations |  from 2018-06-04 to 2020-03-03 |      31
Paypal Europe S A R L Et Cie S C A Ech             |        -13 |      4 operations |  from 2018-05-29 to 2018-10-30 |      32
Vr Permanent Bourso                                |        -60 |     22 operations |  from 2018-06-04 to 2020-03-02 |      31
Vr Permanent Loyer Douillet                        |       -913 |     11 operations |  from 2018-06-04 to 2019-04-03 |      31"
`;

exports[`recurrence bundles should find new bundles 2`] = `"Sepa Mlle Durand Genevieve                         |        -10 |     12 operations |  from 2020-11-30 to 2021-05-31 |       0"`;

exports[`recurrence bundles should update existing bundles 1`] = `
"Cozy Cloud                                         |       3963 |      3 operations |  from 2018-06-01 to 2018-09-28 |    59.5
Edf Clients Particuliers                           |        -36 |      9 operations |  from 2018-08-24 to 2019-04-24 |      31
Free Telecom                                       |        -31 |     12 operations |  from 2018-05-07 to 2019-04-05 |      30
Navigo Annuel Gie Comutitres                       |        -75 |     10 operations |  from 2018-06-04 to 2019-04-03 |      31
Paypal Europe S A R L Et Cie S C A Ech             |        -13 |      4 operations |  from 2018-05-29 to 2018-10-30 |      32
Vr Permanent Bourso                                |        -60 |     11 operations |  from 2018-06-04 to 2019-04-02 |    30.5
Vr Permanent Loyer Douillet                        |       -913 |     11 operations |  from 2018-06-04 to 2019-04-03 |      31"
`;

exports[`recurrence bundles should update existing bundles 2`] = `
"Cozy Cloud                                         |       3963 |     13 operations |  from 2018-06-01 to 2020-02-28 |    31.5
Edf Clients Particuliers                           |        -36 |     11 operations |  from 2018-08-24 to 2019-06-24 |      31
Free Telecom                                       |        -31 |     25 operations |  from 2018-05-07 to 2020-03-06 |      30
Navigo Annuel Gie Comutitres                       |        -75 |     20 operations |  from 2018-06-04 to 2020-03-03 |      31
Paypal Europe S A R L Et Cie S C A Ech             |        -13 |      8 operations |  from 2018-05-29 to 2019-06-11 |      20
Vr Permanent Bourso                                |        -60 |     22 operations |  from 2018-06-04 to 2020-03-02 |      31
Vr Permanent Loyer Douillet                        |       -913 |     11 operations |  from 2018-06-04 to 2019-04-03 |    30.5"
`;
